web/courses.go

```diff
@@ -311,81 +313,32 @@ func (s *AutograderService) changeCourseVisibility(enrollment *pb.Enrollment) er
-func (s *AutograderService) getRepositoryURL(currentUser *pb.User, courseID uint64, repoType pb.Repository_Type) (string, error) {
-	course, err := s.db.GetCourse(courseID, false)
-	if err != nil {
-		return "", err
-	}
-	userRepoQuery := &pb.Repository{
-		OrganizationID: course.GetOrganizationID(),
-		RepoType:       repoType,
-	}
-
-	switch repoType {
-	case pb.Repository_USER:
-		userRepoQuery.UserID = currentUser.GetID()
-	case pb.Repository_GROUP:
-		enrol, err := s.db.GetEnrollmentByCourseAndUser(courseID, currentUser.GetID())
-		if err != nil {
-			return "", err
-		}
-		if enrol.GetGroupID() > 0 {
-			userRepoQuery.GroupID = enrol.GroupID
-		}
-	}
-
-	repos, err := s.db.GetRepositories(userRepoQuery)
-	if err != nil {
-		return "", err
-	}
-	if len(repos) != 1 {
-		return "", fmt.Errorf("found %d repositories for query %+v", len(repos), userRepoQuery)
-	}
-	return repos[0].HTMLURL, nil
-}
-
-// isEmptyRepo returns nil if all repositories for the given course and student or group are empty,
-// returns an error otherwise.
-func (s *AutograderService) isEmptyRepo(ctx context.Context, sc scm.SCM, request *pb.RepositoryRequest) error {
-	course, err := s.db.GetCourse(request.GetCourseID(), false)
-	if err != nil {
-		return err
-	}
-	repos, err := s.db.GetRepositories(&pb.Repository{OrganizationID: course.GetOrganizationID(), UserID: request.GetUserID(), GroupID: request.GetGroupID()})
-	if err != nil {
-		return err
-	}
-	if len(repos) < 1 {
-		return fmt.Errorf("no repositories found")
-	}
-	return isEmpty(ctx, sc, repos)
-}
-
 // rejectEnrollment rejects a student enrollment, if a student repo exists for the given course, removes it from the SCM and database.
 func (s *AutograderService) rejectEnrollment(ctx context.Context, sc scm.SCM, enrolled *pb.Enrollment) error {
 	// course and user are both preloaded, no need to query the database
 	course, user := enrolled.GetCourse(), enrolled.GetUser()
-	repos, err := s.db.GetRepositories(&pb.Repository{
-		UserID:         user.GetID(),
-		OrganizationID: course.GetOrganizationID(),
-		RepoType:       pb.Repository_USER,
-	})
-	if err != nil {
-		return err
+	if err := s.db.RejectEnrollment(user.ID, course.ID); err != nil {
+		s.logger.Debugf("Failed to delete enrollment from database: %v", err)
+		// continue with other delete operations
 	}
-	for _, repo := range repos {
-		// we do not care about errors here, even if the github repo does not exists,
-		// log the error and go on with deleting database entries
-		if err := removeUserFromCourse(ctx, sc, user.GetLogin(), repo); err != nil {
-			s.logger.Debug("updateEnrollment: rejectUserFromCourse failed (expected behavior): ", err)
-		}
-
-		if err := s.db.DeleteRepositoryByRemoteID(repo.GetRepositoryID()); err != nil {
-			return err
-		}
+	repo, err := s.getRepo(course, user.GetID(), pb.Repository_USER)
+	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+		return err // unexpected error
```

Comment:
Should probably add context to the return error; check if callers log errors with context. Other Should include context (user/course) in err message.